### PR TITLE
ZCS-12188: Don't disable zextras zimlets when upgrading a mailbox in …

### PR DESCRIPTION
**Pull request by vishnukantjangid**
**Issue:** Zextras Zimlets are getting loaded in User WebClient and also in Admin console under Zimlet list when upgrading a mailbox in rolling upgrade setup. 

**Solution:** We have added a Constant list that have all the Zextras Zimlets, so while loading the Zimlets we are excluding the Zimlets from the Constant list.

### CODE REVIEW

File: store/src/java/com/zimbra/cs/zimlet/ZimletUtil.java [hunk 1]
```diff
@@ -195,24 +195,48 @@ public static ZimletPresence getAvailableZimlets(Account acct) throws ServiceExc
         if (domain != null) {
             String[] domainZimlets = domain.getMultiAttr(Provisioning.A_zimbraZimletDomainAvailableZimlets);
             for (String zimletWithPrefix : domainZimlets) {
-                availZimlets.put(zimletWithPrefix);
+                if(isNotZextrasZimlet(zimletWithPrefix)) {
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/cs/zimlet/ZimletUtil.java:198]: add space after if

File: store/src/java/com/zimbra/cs/zimlet/ZimletUtil.java [hunk 2]
```diff
@@ -195,24 +195,48 @@ public static ZimletPresence getAvailableZimlets(Account acct) throws ServiceExc
...
+                    availZimlets.put(zimletWithPrefix);
+                }
             }
         }
 
         String[] acctCosZimlets = acct.getMultiAttr(Provisioning.A_zimbraZimletAvailableZimlets);
         for (String zimletWithPrefix : acctCosZimlets) {
-            availZimlets.put(zimletWithPrefix);
+            if(isNotZextrasZimlet(zimletWithPrefix)) {
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/cs/zimlet/ZimletUtil.java:206]: add space after if

File: store/src/java/com/zimbra/cs/zimlet/ZimletUtil.java [hunk 3]
```diff
@@ -195,24 +195,48 @@ public static ZimletPresence getAvailableZimlets(Account acct) throws ServiceExc
...
+        return true;
+    }
 
     public static ZimletPresence getAvailableZimlets(Cos cos) throws ServiceException {
         ZimletPresence availZimlets = new ZimletPresence();
 
         String[] acctCosZimlets = cos.getMultiAttr(Provisioning.A_zimbraZimletAvailableZimlets);
         for (String zimletWithPrefix : acctCosZimlets) {
-            availZimlets.put(zimletWithPrefix);
+            if(isNotZextrasZimlet(zimletWithPrefix)) {
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/cs/zimlet/ZimletUtil.java:237]: add space after if

File: store/src/java/com/zimbra/cs/zimlet/ZimletUtil.java [hunk 4]
```diff
@@ -440,7 +464,9 @@ private static void loadZimletsFromDir(Map<String,ZimletFile> zimlets, String di
                     continue;
                 }
                 try {
-                    zimlets.put(zimletNames[i], new ZimletFile(zimletRootDir, zimletNames[i]));
+                    if(isNotZextrasZimlet(zimletNames[i])) {
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/cs/zimlet/ZimletUtil.java:467]: add space after if

File: store/src/java/com/zimbra/cs/zimlet/ZimletUtil.java [hunk 5]
```diff
@@ -195,24 +195,48 @@ public static ZimletPresence getAvailableZimlets(Account acct) throws ServiceExc
...
-            availZimlets.put(zimletWithPrefix);
+            if(isNotZextrasZimlet(zimletWithPrefix)) {
+                availZimlets.put(zimletWithPrefix);
+            }
         }
 
         return availZimlets;
     }
+    
+    private static boolean isNotZextrasZimlet(String zimletName) {
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/cs/zimlet/ZimletUtil.java:214]: could you  add some doc here for understanding what this method is actually doing with some example

File: store/src/java/com/zimbra/cs/zimlet/ZimletUtil.java (hunk 1 above)
Comment (vishnukantjangid) [reply store/src/java/com/zimbra/cs/zimlet/ZimletUtil.java]: done

File: store/src/java/com/zimbra/cs/zimlet/ZimletUtil.java (hunk 2 above)
Comment (vishnukantjangid) [reply store/src/java/com/zimbra/cs/zimlet/ZimletUtil.java]: done

File: store/src/java/com/zimbra/cs/zimlet/ZimletUtil.java (hunk 3 above)
Comment (vishnukantjangid) [reply store/src/java/com/zimbra/cs/zimlet/ZimletUtil.java]: done

File: store/src/java/com/zimbra/cs/zimlet/ZimletUtil.java (hunk 4 above)
Comment (vishnukantjangid) [reply store/src/java/com/zimbra/cs/zimlet/ZimletUtil.java]: done

**Comment by silentsakky** [comment]
I think commit message is little unclear, explain what exactly is done and why

File: store/src/java/com/zimbra/cs/zimlet/ZimletUtil.java [hunk 6]
```diff
@@ -195,24 +195,57 @@ public static ZimletPresence getAvailableZimlets(Account acct) throws ServiceExc
...
+    /**
+     * this method is also check whether zimletName is having prefix like +,- and !
+     * and this method is used to find out whether zimletName is related to remove
+     * ZEXTRAS list or not if zimletName is not in remove ZEXTRAS list then it
+     * returns true else it returns false
+     *
+     * @param zimletName
+     * @return boolean
+     */
+    private static boolean isNotZextrasZimlet(String zimletName) {
```
Comment (RakeshAMore) [inline store/src/java/com/zimbra/cs/zimlet/ZimletUtil.java:223]: What if we make it `isZextrasZimlet()` and return accordingly?

File: store/src/java/com/zimbra/cs/service/account/GetInfo.java [hunk 7]
```diff
@@ -374,6 +375,7 @@ private static void doZimlets(Element response, Account acct) {
 
             ZimletPresence userZimlets = ZimletUtil.getUserZimlets(acct);
             List<Zimlet> zimletList = ZimletUtil.orderZimletsByPriority(userZimlets.getZimletNamesAsArray());
+            zimletList.removeIf(x -> AdminConstants.ZEXTRAS_PACKAGES_LIST.contains(x.getName()));
```
Comment (RakeshAMore) [inline store/src/java/com/zimbra/cs/service/account/GetInfo.java:378]: may use some meaningful variable name than "x"

**Review by RakeshAMore — APPROVED** [review]

File: store/src/java/com/zimbra/cs/service/account/GetInfo.java (hunk 7 above)
Comment (vishnukantjangid) [reply store/src/java/com/zimbra/cs/service/account/GetInfo.java]: renamed

File: store/src/java/com/zimbra/cs/zimlet/ZimletUtil.java (hunk 6 above)
Comment (vishnukantjangid) [reply store/src/java/com/zimbra/cs/zimlet/ZimletUtil.java]: renamed

**Review by dasiyogesh — APPROVED** [review]

**Review by zimsuchitgupta — APPROVED** [review]

**Review by silentsakky — APPROVED** [review]

**Review by RakeshAMore — APPROVED** [review]

**Pull request merged**